### PR TITLE
Fused Linear and Cross Entropy operations

### PR DESCRIPTION
**Pull request by Jianbing-D**
# Fused Linear Cross Entropy Implementation

This PR introduces an efficient kernel implementation for fused linear (matrix multiplication) and cross entropy operations using Triton. The implementation provides both forward and backward passes with memory and computational optimizations.

## Key Features

- Fuses matrix multiplication with cross entropy computation
- Implements both entropy and cross-entropy calculation in a single pass
- Provides memory-efficient backward pass with two implementation strategies:
  - Total Fused MN: Fuses d_logits & d_hidden & d_weight computation with no intermediate storage (requires FP32 as gradient dtype)
  - Total Separate: Uses intermediate d_logits storage with no special precision requirements

### Equivalent Torch Logic
```python
def run_torch_entropy(hidden: torch.Tensor,
                    weight: torch.Tensor,
                    labels: torch.Tensor) -> typing.List[torch.Tensor]:
    logits = torch.matmul(hidden.to(torch.float32), weight.to(torch.float32)) # [num_tokens, vocab_size]
    pd = torch.nn.functional.softmax(logits, dim=-1) # [num_tokens, vocab_size]
    entropy_a = torch.logsumexp(logits, dim=-1) # [num_tokens]
    entropy_b = torch.sum(pd * logits, dim=-1) # [num_tokens]
    entropy = entropy_a - entropy_b
    logprobs = torch.nn.functional.cross_entropy(logits, labels) # [1]
    return logprobs, entropy
```

### Forward Pass
- Calculates both cross-entropy loss and per-token entropy values
- Fuse multiple operations into fewer kernels to reduce peak memory pressure
- Implements numerically stable softmax

### Backward Pass
- Optimizes memory usage through careful buffer management
- Fuse backward pass into fewer kernels for reduced peak memory pressure and better compute efficiency

## Testing
The implementation includes comprehensive tests that:
- Verify numerical correctness against PyTorch reference implementation
- Compare memory usage between kernel and reference implementations
- Measure and compare forward/backward pass latencies

## Usage Example
```python
hidden = torch.randn(num_tokens, hidden_size, dtype=torch.bfloat16, device="cuda")
weight = torch.randn(hidden_size, vocab_size, dtype=torch.bfloat16, device="cuda")
labels = torch.randint(0, vocab_size, (num_tokens,), device="cuda")

loss, entropy = linear_cross_entropy(hidden, weight, labels, reduction="mean")
```

### CODE REVIEW

**Comment by Jianbing-D** [comment]
## Tested with linting
image

## Tested with pytest
image

**Comment by ksivaman** [comment]
@Jianbing-D Could you sign off your commits? ([instructions](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst#sign-your-work))

**Comment by Jianbing-D** [comment]
> @Jianbing-D Could you sign off your commits? ([instructions](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst#sign-your-work))

Hi, signed off my commits, and force pushed this branch. So that this PR was automatically closed, and a new one was created accordingly. 
https://github.com/NVIDIA/TransformerEngine/pull/1537
